### PR TITLE
[Model] Add DeepGrove Maple (MapleForCausalLM)

### PR DESCRIPTION
**Pull request by NancyFyong**
## Purpose

Adds support for [`deepgrove/maple-preview`](https://huggingface.co/deepgrove/maple-preview) (`MapleForCausalLM`), a 20B-A1B MoE reasoning model (MIT).

Architecture specifics, none of which an existing model file combines:

- `layer_types` interleaves sliding-window and global attention 3:1. The checkpoint runs FlashAttention with `window_size=(sliding_window, 0)`, i.e. an inclusive boundary spanning 513 positions, while vLLM maps `sliding_window=N` to `(N - 1, 0)`; the model therefore passes `config.sliding_window + 1`, as `modernbert.py` already does for the same reason. The window is applied per layer via `per_layer_sliding_window`, and `MapleModel.__init__` clears `cache_config.sliding_window` so the global layers do not inherit it.
- `nope_on_global_attention`: the `full_attention` layers get no positional encoding, so `rotary_emb` is built only for the sliding layers. RoPE is half-width (`partial_rotary_factor: 0.5` → `rotary_dim=64` of `head_dim=128`).
- Per-head QK RMSNorm over the full head dim, before RoPE.
- Experts clamp gate/up to ±7, i.e. `SiluAndMulWithClamp` → `FusedMoEFactory(activation="silu", swiglu_limit=...)`, not `swigluoai` (that adds `up + 1` and `alpha=1.702`).
- `router_dtype: "fp32"` → gate is `ReplicatedLinear(params_dtype=torch.float32)` with `router_logits_dtype=torch.float32`.

### Why `transformers_utils/config.py` is touched

The checkpoint declares `auto_map` but **never serializes `model_type`** (it only exists on the class in `configuration_maple.py`). `_CONFIG_REGISTRY` is keyed by `model_type`, so the lookup is skipped entirely and `AutoConfig` demands `trust_remote_code=True`. Registering the class with Transformers does not help either — `has_local_code` is `"model_type" in config_dict and ... in CONFIG_MAPPING`, and the first clause short-circuits on the missing key.

So the model type is recovered from `architectures`, and the registered class is built directly by reusing the branch that already exists for `_SPECULATIVE_DECODING_CONFIGS`. This keeps `--trust-remote-code` unnecessary and lets vLLM own defaults the checkpoint omits (notably `swiglu_limit`).

### No new parsers

The chat template is Qwen3-shaped, so existing parsers work as-is: `--reasoning-parser qwen3 --tool-call-parser hermes --enable-auto-tool-choice`.

## Test Plan

```bash

pytest tests/models/test_maple.py tests/transformers_utils/test_missing_model_type.py
pytest "tests/models/test_initialization.py::test_can_initialize_large_subset[MapleForCausalLM]"
pytest tests/models/test_registry.py

# 2x H20, bf16
vllm serve deepgrove/maple-preview --max-model-len 8192 [--tensor-parallel-size 2] [--enable-expert-parallel]
python tests/evals/gsm8k/gsm8k_eval.py --port 8000 --num-questions 200 --num-shots 5
```

GSM8K was also run in the model's intended chat/thinking mode (`temperature=0`, answer taken after `</think>`).

Numeric parity against the checkpoint's own modeling code, with only its FA3 kernel swapped for an SDPA mask that reproduces `window_size=(w, 0)` exactly, on a 900-token natural-text prompt (`prompt_logprobs=20` vs. `log_softmax` of the reference logits, aligned so that vLLM position `i` is compared against reference position `i - 1`).

## Test Result

Lint: all hooks pass; mypy passes on 3.10–3.13. `actionlint` skipped (downloads a Go toolchain, and the diff has no workflow files); `--all-files` unavailable because `pre-commit` needs `git ls-files --deduplicate` (git ≥ 2.31, this box has 2.29.2), hence the diff range.

| Test | Result |
| --- | --- |
| `tests/models/test_maple.py` | 8 passed |
| `tests/transformers_utils/test_missing_model_type.py` | 1 passed (verified red when the `architectures` recovery is reverted) |
| `test_can_initialize_large_subset[MapleForCausalLM]` | passed |
| `tests/models/test_registry.py` | 378 passed, 15 skipped (2 pre-existing failures need `timm`, unrelated) |

Weight loading: 9 shards / 37.69 GiB, no missing or unexpected keys.

Parity, top-1 agreement in the windowed region (positions > 513) of the 900-token prompt:

| | vs reference|
| --- | --- |
| vLLM  (this PR) | **93.78%** (362/386) |
| reference vs itself, window shrunk by one token | 93.54% |

93.54% is the ceiling this comparison can reach, so the `+ 1` form is within noise of the reference while the unchanged form is measurably off. Mean |Δ logprob| is 0.091 and next-token top-1 accuracy matches the reference (0.5061 vs 0.5028), confirming there is no systematic divergence.

Two findings worth flagging from that harness:

- The published modeling code is **silently broken under Transformers 5**: `from_pretrained` leaves the non-persistent `inv_freq` buffer uninitialised (observed values `0.0, 0.0, 2.67e18`), so RoPE degenerates to `cos == 1` everywhere and next-token accuracy collapses from 0.50 to 0.04. vLLM builds its own rotary cache and is unaffected; the numbers above were obtained after restoring `inv_freq` in the reference.
- Its RoPE init also fails outright on Transformers 5, since `ROPE_INIT_FUNCTIONS` no longer has a `"default"` entry. Another reason to prefer the in-tree `MapleConfig` over `--trust-remote-code`.

| GSM8K | chat/thinking, 200 q | 5-shot completion, 200 q | output tok/s |
| --- | --- | --- | --- |
| TP=1 | **91.0%** | 35.0% | 5020 |
| TP=2 | **91.0%** (100 q) | — | — |
| TP=2 + EP (128/256 experts per rank) | **90.0%** | 35.5% | 6365 |

The low completion-style score is expected rather than a regression: the template opens `<think>` for the assistant, so plain few-shot continuation is not this model's mode. Generations are coherent under both eager and `torch.compile` + full CUDA graphs. Parsers verified end-to-end: `reasoning` and `content` split correctly, and `tool_calls` returns `{"name": "get_weather", "arguments": "{\"city\": \"Paris\"}"}`.

### Follow-up validation (`a006ca53d6`)

This follow-up adds a CPU regression test for the interleaved-cache failure mode
seen in earlier sliding-window model fixes: a `full_attention` layer with no
per-layer window must not fall back to a model-level cache window.

```bash
python -m pytest -v tests/models/test_maple.py \
  tests/transformers_utils/test_missing_model_type.py \
  'tests/models/test_registry.py::test_registry_imports[MapleForCausalLM]'
CUDA_VISIBLE_DEVICES='' VLLM_TARGET_DEVICE=cpu \
python -m pytest -v tests/transformers_utils/test_missing_model_type.py
python -m pytest -v \
  'tests/models/test_initialization.py::test_can_initialize_large_subset[MapleForCausalLM]'
```

Results:

- Maple unit/config/registry selection: **13 passed**. The new test drives
  `EngineArgs.create_engine_config()` from a config without `model_type` and
  verifies the recovered 3:1 `layer_types` plus
  `cache_config.sliding_window is None`.
- CPU-only config tests: **2 passed**.
- Dummy-weight initialization: passed.
- Real `deepgrove/maple-preview` checkpoint on H20: bf16 TP=1 eager serving,
  prefix caching, default `torch.compile` with full/piecewise CUDA graphs, and
  repeated identical-prefix requests all generated successfully. The compiled
  run captured every configured CUDA graph size and produced identical greedy
  tokens for the repeated request.
- Real checkpoint TP=2 with `--enable-expert-parallel`: passed. Each rank held
  128 of 256 experts and generated eight tokens successfully.
- Spot checks against the original safetensors found exact equality for the
  embedding, Q/K/V and output projections, QK norms, decoder norms, and LM
  head. A standalone MoE check against the checkpoint's fp32 router and
  clamped SwiGLU reference had max/mean absolute errors of 0.015625/0.000626.

## Known limitations

- Speculative decoding: `ngram` works (verified). EAGLE3/DFlash/DSpark do not, because `MapleForCausalLM` does not implement `SupportsEagle3` and no draft checkpoint exists for this target; adding the `EagleModelMixin` hooks is left for a follow-up.
- `num_shared_experts != 0`, `moe_router_enable_expert_bias` and non-`silu` `hidden_act` raise rather than load something subtly wrong.
- `MixtureOfExperts` (EPLB) is not implemented; only `enable_eplb` / `num_redundant_experts` are plumbed through.
- Only tested on H20 (Hopper).
- Parity is measured at the logits level over one prompt, not per layer.

## Accountability

Duplicate-work checks found no existing PR or issue for this model or architecture (`gh pr list`/`gh issue list` over `Maple`, `deepgrove`, `MapleForCausalLM`; the only hits were fuzzy matches on unrelated PRs #42657, #44475, #44633).

AI assistance was used to write and test this change. The human submitter has reviewed every changed line, understands the change end-to-end, and ran the tests and evals reported above.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by NancyFyong** [comment]
Hi @DarkLight1337 @ywang96 @AndreasKaratzas, can you take a look
